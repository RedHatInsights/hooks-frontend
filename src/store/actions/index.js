import _ from 'lodash';
import ApiClient from 'Utilities/notificationsBackendAPI';

export const FETCH_FILTER    = 'FETCH_FILTER';
export const FETCH_ENDPOINTS = 'FETCH_ENDPOINTS';
export const FETCH_ENDPOINT  = 'FETCH_ENDPOINT';
export const SUBMIT_ENDPOINT = 'SUBMIT_ENDPOINT';
export const DELETE_ENDPOINT = 'DELETE_ENDPOINT';
export const NEW_ENDPOINT    = 'NEW_ENDPOINT';
export const TEST_ENDPOINT   = 'TEST_ENDPOINT';
export const FETCH_APPS      = 'FETCH_APPS';

export const fetchEndpoints = (limit, offset, sortBy = 'name asc') => {
    const params = { limit, offset, sort_by: sortBy };
    const query = _.map(params, (value, key) => value ? `${ key }=${ value }` : '');
    const url = `/endpoints?${ query.join('&') }`;

    return {
        type: FETCH_ENDPOINTS,
        payload: ApiClient.get(url),
        meta: {
            endpoint: url,
            sortBy,
            notifications: {
                rejected: {
                    variant: 'danger',
                    title: 'Failed to load endpoints'
                }
            }
        }
    };
};

export const fetchEndpoint = (id) => ({
    type: FETCH_ENDPOINT,
    payload: ApiClient.get(`/endpoints/${ id }`),
    meta: {
        notifications: {
            rejected: {
                variant: 'danger',
                title: `Failed to load endpoint ${ id }`
            }
        }
    }
});

export const createEndpoint = (data) => {
    return {
        type: SUBMIT_ENDPOINT,
        payload: ApiClient.create('/endpoints', { endpoint: data }),
        meta: {
            data,
            notifications: {
                rejected: {
                    variant: 'danger',
                    title: `Failed to create endpoint ${ data.name }`
                },
                fulfilled: {
                    variant: 'success',
                    title: `Endpoint ${ data.name } created`
                }
            }
        }
    };
};

export const updateEndpoint = (id, data) => {
    return {
        type: SUBMIT_ENDPOINT,
        payload: ApiClient.update(`/endpoints/${ id }`, { endpoint: data }),
        meta: {
            data,
            notifications: {
                rejected: {
                    variant: 'danger',
                    title: `Failed to update endpoint ${ data.name }`
                },
                fulfilled: {
                    variant: 'success',
                    title: `Endpoint ${ data.name } updated`
                }
            }
        }
    };
};

export const toggleEndpoint = (id, on) => {
    return {
        type: SUBMIT_ENDPOINT,
        payload: ApiClient.update(`/endpoints/${ id }`, { endpoint: { active: on }}),
        meta: {
            notifications: {
                rejected: {
                    variant: 'danger',
                    title: `Failed to toggle endpoint ${ id }`
                }
            }
        }
    };
};

export const deleteEndpoint = (id, name) => ({
    type: DELETE_ENDPOINT,
    payload: ApiClient.destroy(`/endpoints/${ id }`).then(() => ({ id })),
    meta: {
        notifications: {
            rejected: {
                variant: 'danger',
                title: `Failed to delete endpoint ${ name }`
            },
            fulfilled: {
                variant: 'success',
                title: `Endpoint ${ name } deleted`
            }
        }
    }
});

export const newEndpoint = () => ({
    type: NEW_ENDPOINT
});

export const testEndpoint = (endpointId) => ({
    type: TEST_ENDPOINT,
    payload: ApiClient.create(`/endpoints/${ endpointId }/test`, {}),
    meta: {
        notifications: {
            rejected: {
                variant: 'warning',
                title: 'Test event delivery failed'
            },
            fulfilled: {
                variant: 'success',
                title: 'Test event delivery successful'
            }
        }
    }
});

export const fetchFilter = (endpointId) => ({
    type: FETCH_FILTER,
    payload: ApiClient.get(`/endpoints/${ endpointId }/filter`),
    meta: {
        notifications: {
            rejected: {
                variant: 'danger',
                title: 'Failed to load filter'
            }
        }
    }
});

export const fetchApps = () => ({
    type: FETCH_APPS,
    payload: ApiClient.get('/apps'),
    meta: {
        notifications: {
            rejected: {
                variant: 'danger',
                title: 'Failed to load apps'
            }
        }
    }
});
