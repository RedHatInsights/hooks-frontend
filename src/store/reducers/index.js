import {
    FETCH_FILTERS,
    FETCH_FILTERS_SUCCESS,
    FETCH_FILTERS_FAILURE,
    FETCH_ENDPOINTS,
    FETCH_ENDPOINTS_SUCCESS,
    FETCH_ENDPOINTS_FAILURE,
    FETCH_ENDPOINT_SUCCESS
} from '../actions/index';

const defaultIntialState = {
    loading: false,
    error: null
};

const initialStateFor = function (reducerName) {
    let initState = Object.assign({}, defaultIntialState);
    initState[reducerName] = [];
    return initState;
};

export const endpointReducer = function(state = initialStateFor('endpoints'), action) {
    switch (action.type) {
        case FETCH_ENDPOINTS:
            return {
                ...state,
                loading: true,
                error: null
            };

        case FETCH_ENDPOINTS_SUCCESS:
            return {
                ...state,
                loading: false,
                endpoints: action.payload.endpoints
            };

        case FETCH_ENDPOINT_SUCCESS:
            return {
                ...state,
                loading: false,
                endpoint: action.payload.endpoint
            };

        case FETCH_ENDPOINTS_FAILURE:
            return {
                ...state,
                loading: false,
                error: action.payload.error,
                endpoints: []
            };

        default:
            return state;
    }
};

export const filterReducer = function(state = initialStateFor('filters'), action) {
    switch (action.type) {
        case FETCH_FILTERS:
            return {
                ...state,
                loading: true,
                error: null
            };

        case FETCH_FILTERS_SUCCESS:
            return {
                ...state,
                loading: false,
                filters: action.payload.filters
            };

        case FETCH_FILTERS_FAILURE:
            return {
                ...state,
                loading: false,
                error: action.payload.error,
                filters: []
            };

        default:
            return state;
    }
};
