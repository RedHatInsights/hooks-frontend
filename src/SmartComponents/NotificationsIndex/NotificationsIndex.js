import React, { Component } from 'react';
import {
    Title,
    Button,
    Bullseye,
    EmptyState,
    EmptyStateIcon,
    EmptyStateBody,
    Pagination,
    PaginationVariant
} from '@patternfly/react-core';
import { Link } from 'react-router-dom';
import { CubesIcon } from '@patternfly/react-icons';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import * as actionCreators from 'Store/actions';
import {
    Skeleton,
    SkeletonSize
} from '@red-hat-insights/insights-frontend-components';
import {
    Table,
    TableHeader,
    TableBody,
    sortable
} from '@patternfly/react-table';
import registryDecorator from '@red-hat-insights/insights-frontend-components/Utilities/Registry';
import debounce from 'lodash/debounce';

import './notifications-index.scss';

import {
    EndpointToggle,
    IndexToolbar,
    LoadingState,
    NotificationActions,
    NotificationsPage,
    StatusIcon
} from 'PresentationalComponents';

@registryDecorator()
export class NotificationsIndex extends Component {
    constructor(props) {
        super(props);
        this.state = {
            page: 1,
            perPage: 10,
            rows: [ ],
            sortBy: {
                index: 0,
                direction: 'asc'
            },
            columns: [
                { title: 'Name', key: 'name', transforms: [ sortable ]},
                'Type',
                { title: 'Path', key: 'url', transforms: [ sortable ]},
                'Status',
                { title: 'Active', key: 'active', transforms: [ sortable ]},
                ''
            ]
        };

        this.onSort = this.onSort.bind(this);
    }

    onSort(_event, index, direction) {
        const column = this.state.columns[index].key;
        this.props.fetchEndpoints(this.state.page, this.state.perPage, `${column} ${direction}`).then(() => {
            this.setState({ sortBy: { index, direction }});
            this.filtersInRowsAndCells();
        });
    }

    componentDidMount() {
        this.refreshData();
    }

    changePage = debounce(() => { this.refreshData(); }, 800);

    onPageChange = (_event, page, shouldDebounce) => {
        this.setState({ page });
        if (shouldDebounce) {
            this.changePage();
        } else {
            this.refreshData(page);
        }
    }

    onPageInput = (event, page) =>
        this.onPageChange(event, page, true)

    onSetPage = (event, page) =>
        event.target.className === 'pf-c-form-control' || this.onPageChange(event, page, false)

    refreshData = (page = this.state.page, perPage = this.state.perPage) => {
        this.props.fetchEndpoints(page, perPage).then(() =>
            this.filtersInRowsAndCells()
        );
    }

    onPerPageSelect = (_event, perPage) => {
        this.setState({ perPage });
        this.refreshData(null, perPage);
    }

    filtersInRowsAndCells = () => {
        const endpoints = Object.values(this.props.endpoints);

        let rows = [];
        if (endpoints.length > 0) {
            rows = endpoints.map(({ id, attributes: { active, name, url }}) => (
                [
                    { title: name },
                    { title: 'HTTP' },
                    { title: url },
                    { title: <StatusIcon key={ `notification_status_${id}` } status={ true } /> },
                    { title: <EndpointToggle key={ `notification_switch_${id}` }
                        id={ parseInt(id) }
                        active={ active }
                        onChange={ (checked) => {
                            this.props.toggleEndpoint(id, checked).then(() => this.filtersInRowsAndCells());
                        } } /> },
                    { title: <NotificationActions key={ `notification_actions_${id}` }
                        endpointId={ parseInt(id) }
                        onDelete={ this.onDelete(id, name) }
                        onTest={ this.onTest(id) } /> }
                ]));
        }

        this.setState({ rows });
    }

    onDelete = (id, name) =>
        event => {
            event.preventDefault();
            this.props.deleteEndpoint(id, name).then(() => this.filtersInRowsAndCells());
        }

    onTest = (id) =>
        event => {
            event.preventDefault();
            this.props.testEndpoint(id);
        }

    noResults = () =>
        <Bullseye>
            <EmptyState>
                <p>
                    <EmptyStateIcon icon={ CubesIcon } />
                </p>
                <Title size="lg">No Endpoins found</Title>
                <EmptyStateBody>
                    There are no endpoints configured yet.
                </EmptyStateBody>
                <Button variant="primary" to={ '/new' } component={ Link } onClick={ this.props.newEndpoint }>New endpoint</Button>
            </EmptyState>
        </Bullseye>

    resultsTable = () => {
        const { perPage, page, rows, columns, sortBy } = this.state;

        return <div>
            <Table aria-label='Notifications list'
                rows={ rows }
                cells={ columns }
                sortBy={ sortBy }
                onSort={ this.onSort }>
                <TableHeader />
                <TableBody />
                <tfoot><tr><td colSpan='6'>
                    <Pagination
                        itemCount={ this.props.total }
                        widgetId="pagination-options-menu-bottom"
                        variant={ PaginationVariant.bottom }
                        perPage={ perPage }
                        page={ page }
                        onSetPage={ this.onSetPage }
                        onPageInput={ this.onPageInput }
                        onPerPageSelect={ this.onPerPageSelect } />
                </td></tr></tfoot>
            </Table>
        </div>;
    }

    render() {
        const placeholder = <Skeleton size={ SkeletonSize.lg } />;

        return (
            <NotificationsPage
                title='Notifications'
                rightBar={ <IndexToolbar onClick={ this.props.newEndpoint }/> }>
                <LoadingState
                    loading={ this.props.loading }
                    placeholder={ placeholder } >
                    { this.state.rows.length > 0 ? this.resultsTable() : this.noResults() }
                </LoadingState>
            </NotificationsPage>
        );
    }
}

NotificationsIndex.propTypes = {
    fetchEndpoints: PropTypes.func.isRequired,
    deleteEndpoint: PropTypes.func.isRequired,
    newEndpoint: PropTypes.func.isRequired,
    toggleEndpoint: PropTypes.func.isRequired,
    testEndpoint: PropTypes.func.isRequired,
    endpoints: PropTypes.object.isRequired,
    error: PropTypes.string,
    loading: PropTypes.bool,
    total: PropTypes.number
};

const mapStateToProps = ({ endpoints: { endpoints, loading, error, total }}) => ({
    endpoints,
    loading,
    error,
    total
});

const mapDispatchToProps = (dispatch) =>
    bindActionCreators({
        fetchEndpoints: actionCreators.fetchEndpoints,
        deleteEndpoint: actionCreators.deleteEndpoint,
        newEndpoint: actionCreators.newEndpoint,
        toggleEndpoint: actionCreators.toggleEndpoint,
        testEndpoint: actionCreators.testEndpoint
    }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(NotificationsIndex);
