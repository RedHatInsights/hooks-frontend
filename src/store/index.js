import ReducerRegistry from '@red-hat-insights/insights-frontend-components/Utilities/ReducerRegistry';
import promiseMiddleware from 'redux-promise-middleware';
import { filterReducer, endpointsReducer } from './reducers';
import { appsReducer } from './reducers/AppsReducer';
import { notifications, notificationsMiddleware } from '@red-hat-insights/insights-frontend-components/components/Notifications';

let registry;

export function init (...middleware) {
    if (registry) {
        throw new Error('store already initialized');
    }

    registry = new ReducerRegistry({}, [
        promiseMiddleware(),
        notificationsMiddleware(),
        ...middleware
    ]);

    registry.register({
        filter: filterReducer,
        endpoints: endpointsReducer,
        apps: appsReducer,
        notifications
    });

    return registry;
}

export function getStore () {
    return registry.getStore();
}

export function register (...args) {
    return registry.register(...args);
}
