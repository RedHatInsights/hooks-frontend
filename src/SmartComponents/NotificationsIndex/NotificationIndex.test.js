import { init } from '../../store';
import logger from 'redux-logger';

import { NotificationsIndex } from './NotificationsIndex';

const testEndpoints = [
    {
        id: 1,
        name: 'TEST Endpoint #1',
        url: 'http://endpoint.com',
        active: true,
        filtersCount: 2
    },
    {
        id: 2,
        name: 'TEST Endpoint #2',
        url: 'http://endpoint2.com',
        active: true,
        filtersCount: 1
    },
    {
        id: 3,
        name: 'TEST Endpoint #3',
        url: 'http://endpoint3.com',
        active: false,
        filtersCount: 4
    }
];

describe('NotificationsIndex', () => {
    const store = init(logger).getStore();
    const defaultProps = {
        fetchFilters: jest.fn(),
        fetchEndpoints: jest.fn(),
        deleteEndpoint: jest.fn(),
        toggleEndpoint: jest.fn(),
        newEndpoint: jest.fn(),
        filters: [],
        endpoints: [],
        store
    };

    it('expect to render an EmptyState by default (with no endpoints)', () => {
        const wrapper = shallow(
            <NotificationsIndex { ...defaultProps } />
        );
        expect(wrapper.find('EmptyState').length).toBe(1);
        expect(wrapper).toMatchSnapshot();
    });

    it('takes endpoints', () => {
        const wrapper = shallow(
            <NotificationsIndex { ...defaultProps } endpoints={ testEndpoints }/>
        );
        expect(wrapper).toMatchSnapshot();
    });
});
