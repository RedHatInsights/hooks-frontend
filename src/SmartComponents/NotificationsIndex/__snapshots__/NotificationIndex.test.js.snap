// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationsIndex expect to render an EmptyState by default (with no endpoints) 1`] = `
<withRouter(NotificationsPage)
  rightBar={
    <IndexToolbar
      onClick={[MockFunction]}
    />
  }
  title="Notifications"
>
  <LoadingState
    placeholder={
      <Skeleton
        size="lg"
      />
    }
  >
    <Bullseye
      className=""
      component="div"
    >
      <EmptyState
        className=""
      >
        <p>
          <EmptyStateIcon
            className=""
            icon={[Function]}
          />
        </p>
        <Title
          size="lg"
        >
          No Endpoins found
        </Title>
        <EmptyStateBody
          className=""
        >
          There are no endpoints configured yet.
        </EmptyStateBody>
        <Button
          aria-label={null}
          className=""
          component={[Function]}
          isActive={false}
          isBlock={false}
          isDisabled={false}
          isFocus={false}
          isHover={false}
          onClick={[MockFunction]}
          to="/new"
          type="button"
          variant="primary"
        >
          New endpoint
        </Button>
      </EmptyState>
    </Bullseye>
  </LoadingState>
</withRouter(NotificationsPage)>
`;

exports[`NotificationsIndex takes endpoints 1`] = `
<withRouter(NotificationsPage)
  rightBar={
    <IndexToolbar
      onClick={[MockFunction]}
    />
  }
  title="Notifications"
>
  <LoadingState
    placeholder={
      <Skeleton
        size="lg"
      />
    }
  >
    <div>
      <Table
        aria-label="Notifications list"
        footer={
          <Pagination
            dropDirection="down"
            itemCount={0}
            onFirstClick={[Function]}
            onLastClick={[Function]}
            onNextClick={[Function]}
            onPageInput={[Function]}
            onPerPageSelect={[Function]}
            onPreviousClick={[Function]}
            onSetPage={[Function]}
            page={1}
            perPage={10}
            perPageOptions={
              Array [
                Object {
                  "title": "10",
                  "value": 10,
                },
                Object {
                  "title": "20",
                  "value": 20,
                },
                Object {
                  "title": "50",
                  "value": 50,
                },
                Object {
                  "title": "100",
                  "value": 100,
                },
              ]
            }
            titles={
              Object {
                "currPage": "Current page",
                "items": "items",
                "itemsPerPage": "Items per page",
                "optionsToggle": "Select",
                "pages": "pages",
                "paginationTitle": "Pagination",
                "toFirstPage": "Go to first page",
                "toLastPage": "Go to last page",
                "toNextPage": "Go to next page",
                "toPreviousPage": "Go to previous page",
              }
            }
            toggleTemplate={[Function]}
            variant="bottom"
            widgetId="pagination-options-menu-bottom"
          />
        }
        header={
          Array [
            "Name",
            "Type",
            "Path",
            "Status",
            "Active",
            "",
          ]
        }
        rows={
          Array [
            Object {
              "cells": Array [
                "Test 1",
                "HTTP",
                "http://whatever.com",
                <StatusIcon
                  status={true}
                />,
                <EndpointToggle
                  active={true}
                  id={36}
                  onChange={[Function]}
                />,
                <withRouter(NotificationActions)
                  endpointId={36}
                  onDelete={[Function]}
                />,
              ],
            },
            Object {
              "cells": Array [
                "Test 2",
                "HTTP",
                "http://whatever1.com",
                <StatusIcon
                  status={true}
                />,
                <EndpointToggle
                  active={true}
                  id={37}
                  onChange={[Function]}
                />,
                <withRouter(NotificationActions)
                  endpointId={37}
                  onDelete={[Function]}
                />,
              ],
            },
            Object {
              "cells": Array [
                "Test 3",
                "HTTP",
                "http://whatever2.com",
                <StatusIcon
                  status={true}
                />,
                <EndpointToggle
                  active={true}
                  id={38}
                  onChange={[Function]}
                />,
                <withRouter(NotificationActions)
                  endpointId={38}
                  onDelete={[Function]}
                />,
              ],
            },
          ]
        }
      >
        <TableHeader />
        <TableBody />
      </Table>
    </div>
  </LoadingState>
</withRouter(NotificationsPage)>
`;
